File: honoka-sdk-jvm/honoka-utils/src/main/java/de/honoka/sdk/util/web/ApiResponse.java
package de.honoka.sdk.util.web;

import cn.hutool.http.HttpStatus;
import cn.hutool.json.JSONUtil;
import lombok.Data;
import lombok.experimental.Accessors;

@Accessors(chain = true)
@Data
public class ApiResponse<T> {
    
    private Integer code;
    
    private Boolean success;
    
    private String msg;
    
    private T data;
    
    private ApiResponse() {}
    
    public static <T1> ApiResponse<T1> of() {
        return new ApiResponse<>();
    }
    
    public static <T1> ApiResponse<T1> success(String msg, T1 data) {
        return new ApiResponse<T1>()
            .setCode(HttpStatus.HTTP_OK)
            .setSuccess(true)
            .setMsg(msg)
            .setData(data);
    }
    
    public static <T1> ApiResponse<T1> success(T1 data) {
        return success(null, data);
    }
    
    public static ApiResponse<Object> success() {
        return success(null);
    }
    
    public static ApiResponse<Object> fail(int httpStatus, String msg) {
        return new ApiResponse<>()
            .setCode(httpStatus)
            .setSuccess(false)
            .setMsg(msg)
            .setData(null);
    }
    
    public static ApiResponse<Object> fail(String msg) {
        return fail(HttpStatus.HTTP_INTERNAL_ERROR, msg);
    }
    
    public String toJsonString() {
        return JSONUtil.toJsonStr(this);
    }
}

File: honoka-sdk-jvm/honoka-kotlin-utils/src/main/java/de/honoka/sdk/util/kotlin/web/ApiResponse.kt
package de.honoka.sdk.util.kotlin.web

import cn.hutool.http.HttpStatus

@Suppress("UNCHECKED_CAST")
data class ApiResponse<T>(

    var code: Int? = null,

    var success: Boolean? = null,

    var msg: String? = null,

    var data: T = null as T,

    var error: Any? = null
) {

    companion object {

        fun empty(): ApiResponse<Any?> = ApiResponse()

        fun <T> of(): ApiResponse<T> = ApiResponse()

        fun <T> success(msg: String?, data: T): ApiResponse<T> = ApiResponse(
            HttpStatus.HTTP_OK, true, msg, data
        )

        fun <T> success(data: T): ApiResponse<T> = success(null, data)

        fun success(): ApiResponse<Any?> = success(null)

        fun fail(code: Int?, msg: String?, error: Any?): ApiResponse<Any?> = ApiResponse(
            code, false, msg, error = error
        )

        fun fail(msg: String? = null, error: Any? = null): ApiResponse<Any?> = fail(
            HttpStatus.HTTP_INTERNAL_ERROR, msg, error
        )
    }
}
